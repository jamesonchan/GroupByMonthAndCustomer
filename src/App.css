table {
  width: 100%;
  border: 1px solid grey;
  text-align: center;
}

th {
  border: 1px solid grey;
}

td {
  border: 1px solid grey;
}

body{
  text-align: center;
}