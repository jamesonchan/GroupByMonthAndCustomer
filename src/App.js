import { useEffect, useMemo, useState } from "react";
import "./App.css";
import TableWIthTransactions from "./components/TableWIthTransactions";
import { fetchTransactionMockRecord } from "./services/transaction.api";

const filterByMonth = (transactions, trans) => {
  return transactions.filter((item) => {
    return (
      new Date(item.createDate).getMonth() ===
      new Date(trans.createDate).getMonth()
    );
  });
};

const filterByCustomerId = (filterByMonthArr, trans) => {
  return filterByMonthArr.filter((item) => {
    return item.customerId === trans.customerId;
  });
};

function App() {
  const [transactions, setTransactions] = useState([]);

  // map strcuture : {july : {customerId :[{}]...}, June : {customerId : [{}]... }}
  const parseTransactions = useMemo(() => {
    const map = new Map();
    const res = [];
    // get months on outer map
    transactions.forEach((trans) => {
      const tempMap = new Map();
      const filterByMonthArr = filterByMonth(transactions, trans);

      // get customerId on inner map
      filterByMonthArr.forEach((trans) => {
        const filterByCustomerIdArr = filterByCustomerId(
          filterByMonthArr,
          trans
        );
        // set inner map categorized by customerId
        tempMap.set(trans.customerId, filterByCustomerIdArr);
        // set outer map categorized by month
        map.set(
          new Date(trans.createDate).toLocaleString("en-US", {
            month: "long",
          }),
          tempMap
        );
      });
    });

    map.forEach((value, key) => {
      res.push({ key, value });
    });

    return res;
  }, [transactions]);

  useEffect(() => {
    fetchTransactionMockRecord().then((data) => {
      setTransactions(data);
    });
  }, []);

  return (
    <div className="App">
      <TableWIthTransactions transactions={parseTransactions} />
    </div>
  );
}

export default App;
