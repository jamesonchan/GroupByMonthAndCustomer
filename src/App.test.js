import { render, screen } from "@testing-library/react";
import App from "./App";
import * as Api from "./services/transaction.api";

// jest.mock("./services/transaction.api");
const mockData = [{ customerId: 1, createDate: 165733232 }];
Api.fetchTransactionMockRecord = () => {
  return new Promise((res, rej) => {
    setTimeout(() => {
      res(mockData);
    }, 1000);
  });
};

describe("App", () => {
  it("should match snapshot", () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("response from mock api", () => {
    expect.assertions(1);
    return Api.fetchTransactionMockRecord().then((data) => {
      expect(data).toBe(mockData);
    });
  });
});
