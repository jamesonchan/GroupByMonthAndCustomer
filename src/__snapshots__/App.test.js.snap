// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h1>
      Loading...
    </h1>
  </div>
</DocumentFragment>
`;
